File: LeetCode/Java Solutions/Delete Nodes And Return Forest/delete.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        Set<Integer> toDeleteSet = new HashSet<>();
        for (int val : to_delete) {
            toDeleteSet.add(val);
        }
        List<TreeNode> remainingForest = new ArrayList<>();
        
        helper(root, true, toDeleteSet, remainingForest);
        return remainingForest;
    }
    
    private TreeNode helper(TreeNode node, boolean isRoot, Set<Integer> toDeleteSet, List<TreeNode> remainingForest) {
        if (node == null) {
            return null;
        }
        boolean rootDeleted = toDeleteSet.contains(node.val);
        if (isRoot && !rootDeleted) {
            remainingForest.add(node);
        }
        node.left = helper(node.left, rootDeleted, toDeleteSet, remainingForest);
        node.right = helper(node.right, rootDeleted, toDeleteSet, remainingForest);
        return rootDeleted ? null : node;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        TreeNode root = new TreeNode(1, 
            new TreeNode(2, new TreeNode(4), new TreeNode(5)), 
            new TreeNode(3, new TreeNode(6), new TreeNode(7))
        );
        List<TreeNode> result = sol.delNodes(root, new int[]{3, 5});
        for (TreeNode tree : result) {
            // print the roots of the remaining forest
            System.out.println(tree.val);
        }
    }
}

File: LeetCode/Python Solutions/Delete Nodes And Return Forest/delete.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def delNodes(self, root: Optional[TreeNode], to_delete: List[int]) -> List[TreeNode]:
        to_delete_set = set(to_delete)
        remaining_forest = []
        
        def helper(node, is_root):
            if not node:
                return None
            root_deleted = node.val in to_delete_set
            if is_root and not root_deleted:
                remaining_forest.append(node)
            node.left = helper(node.left, root_deleted)
            node.right = helper(node.right, root_deleted)
            return None if root_deleted else node
        
        helper(root, True)
        return remaining_forest

# Example usage:
root = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3, TreeNode(6), TreeNode(7)))
sol = Solution()
result = sol.delNodes(root, [3, 5])
for tree in result:
    # print the roots of the remaining forest
    print(tree.val)
